custom/icds/models.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.functional import cached_property
+from django.utils.translation import ugettext_lazy as _
+
+from corehq.apps.app_manager.dbaccessors import get_build_by_version
+from corehq.motech.utils import b64_aes_decrypt
+from custom.icds.utils.ccz_hosting import CCZHostingUtility
+from custom.icds.validators import (
+    LowercaseAlphanumbericValidator,
+)
+
+
+class CCZHostingLink(models.Model):
+    identifier = models.CharField(null=False, unique=True, max_length=255, db_index=True,
+                                  validators=[LowercaseAlphanumbericValidator])
+    username = models.CharField(null=False, max_length=255)
+    # b64_aes_encrypt'ed raw password is stored in DB
+    password = models.CharField(null=False, max_length=255)
+    domain = models.CharField(null=False, max_length=255)
+    page_title = models.CharField(blank=True, max_length=255)
+
+    def __str__(self):
+        return self.identifier
+
+    @cached_property
+    def get_password(self):
+        return b64_aes_decrypt(self.password)
+
+    def to_json(self):
+        from custom.icds.serializers import CCZHostingLinkSerializer
+        return CCZHostingLinkSerializer(self).data
+
+    def delete(self, *args, **kwargs):
+        for ccz_hosting in CCZHosting.objects.filter(link=self):
+            ccz_hosting.delete()
+        super(CCZHostingLink, self).delete(*args, **kwargs)
+
+
+class CCZHosting(models.Model):
+    link = models.ForeignKey(CCZHostingLink, on_delete=models.CASCADE)
+    app_id = models.CharField(max_length=255, null=False)
+    version = models.IntegerField(null=False)
+    profile_id = models.CharField(max_length=255, blank=True)
+    file_name = models.CharField(max_length=255, blank=True)
+
+    class Meta:
+        unique_together = ('link', 'app_id', 'version', 'profile_id')
+
+    @cached_property
+    def utility(self):
+        return CCZHostingUtility(self)
+
+    @cached_property
+    def domain(self):
+        return self.link.domain
+
+    def to_json(self, app_names):
+        from custom.icds.serializers import CCZHostingSerializer
+        return CCZHostingSerializer(self, context={'app_names': app_names}).data
+
+    @cached_property
+    def blob_id(self):
+        assert self.app_id
+        assert self.version
+        return "%s%s%s" % (self.app_id, self.version, self.profile_id)
```

Comment:
I'd have expected some kind of delimiter, like `"%s-%s-%s"`, both for readability while debugging and for the very unlikely case that app_id+version+profile_id. Or is this style typical of blob ids?

Reply:
thats not a typical style of blob ids. This is generated specifically for this only.
I dont think one should ever need to use this to deduce the app_id, version and profile_id since that can be seen on the object itself. I am fine with keeping this as a string of characters that look like an id.

Reply:
okay, that sounds fine

Reply:
Not a big deal, but `"%s-%s-%s"` would be a fine format for such a long and complex multipart blob id.